src/renderer/states/load.ts

```diff
@@ -1,136 +1,91 @@
-  private loader : Phaser.Text
-  private files : Phaser.Text
-  private file : Phaser.Text
+const store = new Store()
 
-  get name(): string {
-    return 'load';
-  }
+// load any external assets and store them in an scoped array
+let external_assets = []
+ipc.on('asset-list', (e, files) => external_assets = files)
 
-  create() {
-    const x = game.world.centerX
-    const y = game.world.centerY
-    const font_style1 = { font: '20px Verdana', fill: '#FFF', align: 'center' }
-    const font_style2 = { font: '12px Verdana', fill: '#FFF', align: 'center' }
-    this.loader = game.add.text(0,y    , `Loading ${0}%`, font_style1)
-    this.files  = game.add.text(0,y+40 , "Files 0 / 0"  , font_style2)
-    this.file   = game.add.text(0,y+70 , ""             , font_style2)
-    this.loader.visible = false
-    this.files.visible  = false
-    this.file.visible   = false
-
-    game.load.onLoadStart.add(this.load_start   , this)
-    game.load.onFileComplete.add(this.file_complete, this)
-    game.load.onLoadComplete.add(this.load_complete, this)
-    this.load_all()
-  }
+/** Loads all sprites, spritesheets, sound effects, etc to the phaser game
+ *  Uses signals to detect once everything has been loaded correctly
+ */
+export default class LoadState extends State {
+  private all_assets : Array<{}>
+  private external_asset_dir : string
 
-  load_start() {
+  get name(): string { return 'load'; }
 
+  /** preload a json file where all assets preconfigs */
+  preload() {
+    game.load.json('assets', '../src/renderer/core/assets.json');
   }
 
-  file_complete =(progress,key,success,cur,total)=> {
-    this.files.setText(`Files ${cur} / ${total}`)
-    this.file.setText(key)
-  }
+  create() {
+    this.external_asset_dir = store.get('asset-dir')
+    this.all_assets = game.cache.getJSON('assets').assets
+    this.load_assets();
 
-  load_complete() {
-    controls.create()
-    game.sounds.create()
-    game.state.start('menu')
-  }
+    game.load.onLoadComplete.add(() => { 
+      controls.create()
+      game.sounds.create()
+      game.state.start('menu')
+    }, this)
 
-  update() {
-    this.loader.setText(`Loading ${game.load.progress}%`);
+    game.load.start();
   }
-
-  render() {
-    //this.loader.visible = true
-    //this.files.visible  = true
-    //this.file.visible   = true
+  
+  /** Loads all assets in the assets folder
+   *  if any assets with the same name, data type etc exist in the 
+   *  external assets folder -> external asset is loaded instead of the
+   *  actual asset.
+   *  
+   *  This allows other to replace any assets with their own if 
+   *  they behave under the restrictions defined in assets.json  
+   */
+  load_assets() {
+    this.all_assets.forEach(e => {
+      let type : string = e["type"];
+      let name : string = e["name"];
+      let load_external = external_assets.find(a => a === name)
+      let path : string = this.asset_path(type, name, load_external)
+
+      switch (type) {
```

Comment:
It makes sense to generalize the imports so you can import anything as a mod.
I wonder if we should be validating what we need to have to run the base game since you can't have certain assets missing.